UniversalAlgebra/Complexity/CSP.lagda

```diff
@@ -15,67 +15,83 @@ module Complexity.CSP where
 
 open import Agda.Primitive         using    ( _⊔_ ; lsuc ; Level)
                                    renaming ( Set to Type )
-open import Data.Nat.Base          using    ( ℕ )
-open import Data.Fin.Base          using    ( Fin )
+open import Function.Base          using    ( _∘_ )
 open import Relation.Unary         using    ( _∈_; Pred   )
 
 open import Relations.Continuous   using    ( Rel )
 
 \end{code}
 
-#### Finite CSP
+#### Constraints
 
-An instance of a (finite) constraint satisfaction problem is a triple 𝑃 = (𝑉, 𝐷, 𝐶) where
+A constraint c consists of
 
-* 𝑉 is a finite set of variables,
-* 𝐷 is a finite domain,
-* 𝐶 is a finite list of constraints, where
-  each constraint is a pair 𝐶 = (𝑥, 𝑅) in which
-  * 𝑥 is a tuple of variables of length 𝑛, called the scope of 𝐶, and
-  * 𝑅 is an 𝑛-ary relation on 𝐷, called the constraint relation of 𝐶.
+ 1. a scope function,  s : I → var, and
 
-\begin{code}
+ 2. a constraint relation, i.e., a predicate over the function type I → D
+
+        I ···> var
+         .     .
+          .   .
+           ⌟ ⌞
+            D
 
-private variable χ ρ : Level
 
-record Constraint (∣V∣ ∣D∣ : ℕ) : Type (lsuc (χ ⊔ ρ)) where
- field
-  scope : Fin ∣V∣ → Type χ -- a subset of Fin ∣V∣
+The *scope* of a constraint is an indexed subset of the set of variable symbols.
+We could define a type for this, e.g.,
 
-  relation : ((v : Fin ∣V∣ ) → v ∈ scope → Fin ∣D∣) → Type ρ
-  -- a subset of "tuples" (i.e., functions 𝑓 : scope → Fin ∣D∣)
+```
+ Scope : Type ν → Type ι → _
+ Scope V I = I → V
+```
 
- -- An assignment 𝑓 : 𝑉 → 𝐷 of values in the domain to variables in 𝑉
- -- *satisfies* the constraint 𝐶 = (𝑥, 𝑅) if 𝑓 ↾ 𝑥 ∈ 𝑅.
- satisfies : (Fin ∣V∣ → Fin ∣D∣) → Type ρ
- satisfies f = (λ v _ → f v) ∈ relation
+but we omit this definition because it's so simple; to reiterate,
+a scope of "arity" I on "variables" V is simply a map from I to V,
+where,
 
+* I denotes the "number" of variables involved in the scope
+* V denotes a collection (type) of "variable symbols"
+
+\begin{code}
 
-open Constraint
+module _ -- levels for...
+         {ι : Level} -- ...arity (or argument index) types
+         {ν : Level} -- ...variable symbol types
+         {δ : Level} -- ... domain types
+         -- {ρ : Level} -- ... relation types
+         where
 
-record CSP : Type (lsuc (χ ⊔ ρ)) where
- field
-  ∣V∣ -- # of variables,
-   ∣D∣ -- # of elements in the domain,
-    ∣C∣ -- # of constraints (resp.)
-       : ℕ
-  𝐶    : Fin ∣C∣ → Constraint{χ} ∣V∣ ∣D∣
+ record Constraint (var : Type ν)(dom : Type δ){ρ : Level} : Type (lsuc (ι ⊔ ν ⊔ δ ⊔ ρ)) where
```

Comment:
Isn't the level that you want `(lsuc ι) ⊔ ν ⊔ δ ⊔ (lsuc ρ)` ? This is lower (in general). The nice thing about parameters is that they don't cause level creep.

Reply:
Correct.  How could you possibly have noticed that?!

Reply:
Experience?